embulk-util-aws-credentials/src/main/java/org/embulk/util/aws/credentials/AwsCredentials.java

```diff
@@ -189,9 +187,6 @@ public void refresh()
 
     private static AWSCredentialsProvider overwriteBasicCredentials(AwsCredentialsConfig task, final AWSCredentials creds)
     {
-        task.setAuthMethod("basic");
-        task.setAccessKeyId(Optional.of(creds.getAWSAccessKeyId()));
-        task.setSecretAccessKey(Optional.of(creds.getAWSSecretKey()));
```

Comment:
The change above is okay because it's only for `profile`. But I was worried that this may affect the existing behavior of `properties` and `env`? (They are calling `overwriteBasicCredentials`, too.)

Reply:
I think this change is safe for `properties` and `env` auths.

### auth_method: properties

`com.amazonaws.auth.SystemPropertiesCredentialsProvider` is called With `auth_method: properties`.
https://github.com/embulk/embulk-input-s3/blob/v0.3.5/embulk-util-aws-credentials/src/main/java/org/embulk/util/aws/credentials/AwsCredentials.java#L130

This class refers to System property directly. Thus, we don't have to set ACCESS_KEY and SECRET_KEY for the plugin tasks.

```java
public class SystemPropertiesCredentialsProvider implements AWSCredentialsProvider {

    @Override
    public AWSCredentials getCredentials() {
        String accessKey = StringUtils.trim(System.getProperty(ACCESS_KEY_SYSTEM_PROPERTY));
        String secretKey = StringUtils.trim(System.getProperty(SECRET_KEY_SYSTEM_PROPERTY));
        String sessionToken = StringUtils.trim(System.getProperty(SESSION_TOKEN_SYSTEM_PROPERTY));
        ...
```

### auth_method: env

`com.amazonaws.auth.EnvironmentVariableCredentialsProvider` is called With `auth_method: env`.
https://github.com/embulk/embulk-input-s3/blob/v0.3.5/embulk-util-aws-credentials/src/main/java/org/embulk/util/aws/credentials/AwsCredentials.java#L95

This class refers to environment variables. Thus, we don't have to set ACCESS_KEY and SECRET_KEY for the plugin tasks as same as `auth_method:env`.

```java
public class EnvironmentVariableCredentialsProvider implements AWSCredentialsProvider {
    @Override
    public AWSCredentials getCredentials() {
        String accessKey = System.getenv(ACCESS_KEY_ENV_VAR);
        if (accessKey == null) {
            accessKey = System.getenv(ALTERNATE_ACCESS_KEY_ENV_VAR);
        }

        String secretKey = System.getenv(SECRET_KEY_ENV_VAR);
        if (secretKey == null) {
            secretKey = System.getenv(ALTERNATE_SECRET_KEY_ENV_VAR);
        }
        ...
```

Reply:
And speaking about `task.setAuthMethod("basic");`, the behavior would not change for both auth methods.

* With calling `task.setAuthMethod("basic")`

  * Called for the 1st time:
    Plugin read from environment variables/properties file
  * Called after the 2nd time:
    Plugin use values in plugin task

* Without calling `task.setAuthMethod("basic")`

  * Called for the 1st time:
    Plugin read from environment variables/properties file
  * Called after the 2nd time:
    Plugin read from environment variables/properties file

Reply:
Got it. Thanks for the detailed explanation!

Then (not mandatory soon, but for the future), we may no longer need the `task` argument in `overwriteBasicCredentials`?

Reply:
Ah, yes. And we may need to refactor this class. I think this class has not been maintained for 2.5 years and contains some redundant codes.